### PR TITLE
chore: Allow datasource to be null in client payloads

### PR DESCRIPTION
**Pull request by nidhi-nair**
Fixes issue with server messing up update action request when client sends a partial payload.

## Summary by CodeRabbit

- **Refactor**
  - Improved the handling of data sources within actions to ensure reliability when creating new actions or collections.

- **Documentation**
  - Updated method documentation to better explain the action validation and saving process.

- **Bug Fixes**
  - Addressed an issue where actions with unspecified data sources could lead to unexpected behavior. Now, a new data source is automatically generated if none is provided.

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/LayoutActionServiceCEImpl.java [hunk 1]
```diff
@@ -466,13 +466,16 @@ public Mono<ActionDTO> createAction(ActionDTO action, AppsmithEventContext event
...
+                                savedActionDTO.getDatasource().getId());
                     } else {
-                        return Mono.justOrEmpty(action.getDatasource());
+                        return Mono.justOrEmpty(savedActionDTO.getDatasource());
                     }
-                }))
+                })
                 .flatMap(zippedData -> {
                     final Tuple2<ActionDTO, NewAction> zippedActions = zippedData.getT1();
                     final Datasource datasource = zippedData.getT2();
```
[the remark this answers is not in the document]
Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/LayoutActionServiceCEImpl.java:481]: Not required, the validateAndSave method make sure that this init is done even if it were null before. That is why the zipWhen change has been done, to make sure we wait for that method to complete.

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/ce/ActionCE_DTO.java [hunk 2]
```diff
@@ -298,4 +282,9 @@ public LayoutExecutableUpdateDTO createLayoutExecutableUpdateDTO() {
 
         return layoutExecutableUpdateDTO;
     }
+
+    public void autoGenerateDatasource() {
+        this.datasource = new Datasource();
```
Comment (subrata71) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/ce/ActionCE_DTO.java:287]: Do you think adding a `null` check before this initialization makes sense here? Imagine calling this function out of mistake will just reset the datasource.

**Review by subrata71 — APPROVED** [review]

**Pull request merged**